tools/numasched.py

```diff
@@ -0,0 +1,141 @@
+#!/usr/bin/env python
+# @lint-avoid-python-3-compatibility-imports
+#
+# numasched  Trace task NUMA switch
+#            For Linux, uses BCC, eBPF.
+#
+# USAGE: numasched [-p PID]
+#
+# This script tracks NUMA migrations of tasks, and in general, frequent
+# NUMA migrations can cause poor performance.
+#
+# Copyright 2022 CESTC, Co.
+# Licensed under the Apache License, Version 2.0 (the "License")
+#
+# 14-Dec-2022   Rong Tao    Created this.
+
+from __future__ import print_function
+from bcc import BPF
+import argparse
+from time import strftime
+from socket import inet_ntop, AF_INET, AF_INET6
+from struct import pack
+from time import sleep
+
+
+# arguments
+examples = """examples:
+    ./numasched             # trace all processes
+    ./numasched -p 185      # trace PID 185 only
+"""
+parser = argparse.ArgumentParser(
+    description="Trace task NUMA switch",
+    formatter_class=argparse.RawDescriptionHelpFormatter,
+    epilog=examples)
+parser.add_argument("-p", "--pid",
+    help="trace this PID only")
+args = parser.parse_args()
+
+pid = args.pid
+
+
+bpf_text = """
+#include <linux/sched.h>
+#include <linux/topology.h>
+
+struct data_t {
+    char comm[TASK_COMM_LEN];
+    u32 pid;
+    u32 old_nid;
+    u32 new_nid;
+};
+BPF_PERF_OUTPUT(events);
+
+struct val_t {
+    char comm[TASK_COMM_LEN];
+    u32 nid;
+};
+BPF_HASH(numaid_info, u32, struct val_t);
+
+static int comm_strcmp(const char *cs, const char *ct)
+{
+    int len = 0;
+    unsigned char c1, c2;
+
+    while (len++ < TASK_COMM_LEN) {
+        c1 = *cs++;
+        c2 = *ct++;
+        if (c1 != c2)
+            return c1 < c2 ? -1 : 1;
+        if (!c1)
+            break;
+    }
+    return 0;
+}
+
+TRACEPOINT_PROBE(sched, sched_switch)
+{
+    u32 pid = bpf_get_current_pid_tgid() >> 32;
+    u32 new_nid = bpf_get_numa_node_id();
+    struct val_t val = {}, *valp;
+    u32 old_nid;
+
+    if (FILTER_PID)
+        return 0;
+
+    bpf_get_current_comm(&val.comm, sizeof(val.comm));
+    val.nid = new_nid;
+
+    valp = numaid_info.lookup(&pid);
+    if (!valp)
+        goto update;
+
+    old_nid = valp->nid;
+
+    if (old_nid != new_nid && !comm_strcmp(val.comm, valp->comm)) {
```

Comment:
any particular reason you compare based comm instead of tid ((u32)bpf_get_current_pid_tgid())?
do we need to handle cases like (one process has two threads, one thread moved to a different numa node while the other stays in the same numa node)?

Reply:
Thanks, Song. When I wrote this tool, I didn't take the multithreaded process model into account, thank you for the review. I just submitted code that uses tid as a judgment condition and added --tid and --comm command line arguments. I have also revised the following modification suggestions, thank you.